TakingITGlobal/commit2act
# [TIG-778] Restructure app routing to enable signup flow

## Ticket(s)

* [TIG-778](https://app.clickup.com/t/9009201449/TIG-778) (focused sub-task)
* [TIG-734](https://app.clickup.com/t/9009201449/TIG-734) (parent task)

## Description

This PR is doing quite a bit of lifting. Enough so that I needed to break out [a sibling task to finish off the user login state and signup flow routing](https://app.clickup.com/t/9009201449/TIG-779) and keep this one somewhat more reasonably-sized. The good news is that _most_ of the 54 files changed are basically just duplicate changes in the component, story, and test files for each signup page.

The basic scope of this PR involves updating the C2A's app routing to allow us to more easily control public vs private vs admin routes throughout the app - while also serving alternate page containers and layout styles between each type. The original app routing assumed that all pages for a logged-in user would render with the same page layout, menu, and theming.  

**Important pieces in this PR:**

* [Conditional routing in `App.js`](https://github.com/TakingITGlobal/commit2act/pull/113/files#diff-3d74dddefb6e35fbffe3c76ec0712d5c416352d9449e2fcc8210a9dee57dff67R93)
* [Conditional rendering of navigation vs state change in `Login_material.js`](https://github.com/TakingITGlobal/commit2act/pull/113/files#diff-4481852e6133ecb7bd1e9a39871a14c2239f82d53ff28d9696a45eb5fa559181R718)
* [Updated login and signup paths in `page-paths.js`](https://github.com/TakingITGlobal/commit2act/pull/113/files#diff-a549b5d62780cbbbf0388da3e9b6aae1e1329a44b28da771fa72feb8de7ed7cfR8)
* [Changes to path names in `signup-paths.js` to keep protected routes under the `/onboard` root](https://github.com/TakingITGlobal/commit2act/pull/113/files#diff-94f16b69b3d48055cefcd72e082359419667b1f4ff89dc595885748dae20758eR3)
* [New `Routing` component introduces a more global-oriented routes setup for the app](https://github.com/TakingITGlobal/commit2act/pull/113/files#diff-70d59857ba4845bfa1a539ef202943acdc2b076c7c2b686a778bc1919f06f90dR1)
* [Our new signup routes have been removed from the `PublicRoutes` file, as they're migrating to the `Routing` handler](https://github.com/TakingITGlobal/commit2act/pull/113/files#diff-185e68dcb4222a07b4feb50c9fc6fb501e93dcd9f8b8c576dfe745ac7dd82936R1)
* [Updating `SignupContainer` to use router `Outlet`s instead of exclusively `children` - allowing us to use it as a wrapper in `Routing`](https://github.com/TakingITGlobal/commit2act/pull/113/files#diff-bdc052dc83aecd13b869825d251bcf11862971511468af84e3d4f93044359989R4)
* [SImilar conditional `Outlet` treatment for `PageContainer`, based on the current signup flow type](https://github.com/TakingITGlobal/commit2act/pull/113/files#diff-d57d9dc9127bc7ac80f0b56b904dee36bc1eec80f87e9cd07f4862281dd9e25fR150)

**The Signup page changes are all basically duplicates, so they don't need as much attention as the rest:**

* [Each of the signup flow's pages have been updated to remove the built-in nesting of `SignupContainer` and ditch the more brittle per-page routing checks for step, progress, etc.](https://github.com/TakingITGlobal/commit2act/pull/113/files#diff-6346b27119e5f70792df1aa995c08b9dd925dfcbf378edc766bf268424632d06R1)

## Testing

1. Pull down the branch and install (`npm install`)
2. Run the test suite (`npm run test`) and confirm everything passes
3. Run the local Storybook instance (`npm run storybook`)
4. Navigate to Pages > Sign Up and confirm each page still renders with expected progress, headings, buttons, etc.
5. Navigate to Views > SignupContainer and confirm the new "With Page as Child Content" story renders the first signup page
6. Run the local app (`npm run start`)
7. From the login page, confirm the "Create account" button changes page state and does not redirect
8. Login with a team or admin user and confirm all expected pages render and menu items appear
9. Logout and shut down the local app (Ctrl + C in Terminal)
10. Change the `REACT_APP_SIGNUP_FLOW` value in the `.env` file to `onboard` from `simple`
11. Rerun the local app (`npm run start`)
12. From the login page, confirm selecting "create account" redirects to the new Sign Up page 
13. Enter a random sub-page as the URL and confirm you are redirected to the login page
14. Login with a team or admin user and confirm all expected pages render and menu items appear

## Screenshots

**Sub-page story added for `SignupContainer`:**

![image](https://github.com/TakingITGlobal/commit2act/assets/1672105/e9d59867-2e0a-40d0-9fba-75e7b007e149)

**Hovering the "create account" button to reveal the URL (bottom left) pointing to `/signup`:**

![image](https://github.com/TakingITGlobal/commit2act/assets/1672105/5307af59-6ed1-4249-a280-85bd4fadd99f)
